File: src/main/java/no/jstien/bikeapi/tsdb/read/RequestFactory.java
package no.jstien.bikeapi.tsdb.read;

import java.time.ZonedDateTime;
import java.util.Arrays;
import java.util.Optional;

public class RequestFactory {
    private ZonedDateTime from;
    private ZonedDateTime to;
    private int[] stationIds;
    private Optional<String> downsample;

    public RequestFactory(ZonedDateTime from, ZonedDateTime to, int ...stationIds) {
        this.from = from;
        this.to = to;
        this.stationIds = stationIds;
    }

    public void setDownsampleMinutes(int minutes) {
        if (minutes <= 0) {
            throw new RuntimeException("Downsample value must be positive");
        }

        downsample = Optional.of(""+minutes+"m-avg");
    }

    Request create() {
        Request request = new Request(from, to);
        Arrays.stream(stationIds).forEach(id -> addQuery(request, id));

        downsample.ifPresent(ds -> {
            request.getQueries().forEach(query -> {
                query.setDownsample(ds);
            });
        });

        return request;
    }

    int[] getStationIds() {
        return stationIds;
    }

    private void addQuery(Request request, int stationId) {
        Query bikeQuery = new Query("bikes.free");
        bikeQuery.addFilter(Filter.literalOrFilter("stationId", false, String.valueOf(stationId)));

        Query lockQuery = new Query("locks.free");
        lockQuery.addFilter(Filter.literalOrFilter("stationId", false, String.valueOf(stationId)));

        request.addQuery(bikeQuery);
        request.addQuery(lockQuery);
    }
}

File: src/main/java/no/jstien/bikeapi/tsdb/read/RequestFactory.kt
package no.jstien.bikeapi.tsdb.read

import java.time.ZonedDateTime
import java.util.Arrays
import java.util.Optional

class RequestFactory(private val from: ZonedDateTime, private val to: ZonedDateTime, vararg stationIds: Int) {
    internal val stationIds: IntArray = stationIds
    private var downsample: Optional<String>? = null

    init {
        this.downsample = Optional.empty()
    }

    fun setDownsampleMinutes(minutes: Int) {
        if (minutes <= 0) {
            throw RuntimeException("Downsample value must be positive")
        }

        downsample = Optional.of("" + minutes + "m-avg")
    }

    internal fun create(): Request {
        val request = Request(from, to)
        Arrays.stream(stationIds).forEach { id -> addQuery(request, id) }

        downsample!!.ifPresent { ds -> request.getQueries().forEach { query -> query.setDownsample(ds) } }

        return request
    }

    private fun addQuery(request: Request, stationId: Int) {
        val bikeQuery = Query("bikes.free")
        bikeQuery.addFilter(Filter.literalOrFilter("stationId", false, stationId.toString()))

        val lockQuery = Query("locks.free")
        lockQuery.addFilter(Filter.literalOrFilter("stationId", false, stationId.toString()))

        request.addQuery(bikeQuery)
        request.addQuery(lockQuery)
    }
}
